### PR TITLE
Refactor: lists and paths in dev

### PR DESCRIPTION
**Pull request by eumiro**

### CODE REVIEW

File: airflow/providers/google/cloud/operators/spanner.py [hunk 1]
```diff
@@ -302,7 +302,7 @@ def sanitize_queries(queries: list[str]) -> None:
         :param queries: queries
         """
         if queries and queries[-1] == "":
-            del queries[-1]
+            queries.pop()
```
Comment (uranusjr) [inline airflow/providers/google/cloud/operators/spanner.py:305]: I’m not sure I like using `pop()` more

File: dev/breeze/src/airflow_breeze/utils/add_back_references.py [hunk 2]
```diff
@@ -52,26 +51,12 @@ def download_file(url):
...
-                continue
-
-            # Skip comments
-            if line.startswith("#"):
-                continue
-
-            line = line.rstrip()
-            file_content.append(line)
-
+    for line in file_name.read_text().splitlines():
```
Comment (uranusjr) [inline dev/breeze/src/airflow_breeze/utils/add_back_references.py]: I would use the `for line in f:` style instead.

File: dev/breeze/src/airflow_breeze/utils/add_back_references.py [hunk 3]
```diff
@@ -119,17 +104,15 @@ def generate_back_references(link: str, base_path: Path):
     old_to_new.append(("index.html", "security.html"))
     old_to_new.append(("security.html", "security/security-model.html"))
 
-    versions = [f.path.split("/")[-1] for f in os.scandir(base_path) if f.is_dir()]
-    for version in versions:
-        print(f"Processing {base_path}, version: {version}")
-        versioned_provider_path = base_path / version
+    versioned_provider_paths = [p for p in base_path.iterdir() if p.is_dir()]
+    for versioned_provider_path in versioned_provider_paths:
```
Comment (uranusjr) [inline dev/breeze/src/airflow_breeze/utils/add_back_references.py]: I think we can avoid building the intermediate list here.

File: dev/breeze/src/airflow_breeze/utils/add_back_references.py [hunk 4]
```diff
@@ -119,17 +104,15 @@ def generate_back_references(link: str, base_path: Path):
...
 
         for old, new in old_to_new:
             # only if old file exists, add the back reference
-            if os.path.exists(versioned_provider_path / old):
-                split_new_path = new.split("/")
-                file_name = new.split("/")[-1]
-                dest_dir = versioned_provider_path.joinpath(*split_new_path[: len(split_new_path) - 1])
+            if (versioned_provider_path / old).exists():
+                *split_new_path, file_name = new.rsplit("/")
+                dest_dir = versioned_provider_path.joinpath(*split_new_path)
```
Comment (uranusjr) [inline dev/breeze/src/airflow_breeze/utils/add_back_references.py]:
```suggestion
                split_new_path, file_name = new.rsplit("/", 1)
                dest_dir = versioned_provider_path.joinpath(split_new_path)
```

Avoid some unnecessary splitting

File: dev/perf/scheduler_dag_execution_timing.py [hunk 5]
```diff
@@ -297,35 +296,30 @@ def main(num_runs, repeat, pre_create_dag_runs, executor_class, dag_ids):
...
     if len(times) > 1:
-        print(
-            (msg + " (±%.3fs)")
-            % (num_runs, len(dags), total_tasks, statistics.mean(times), statistics.stdev(times))
-        )
+        msg += f"{statistics.mean(times):.4f}s (±{statistics.stdev(times):.3f}s)"
     else:
-        print(msg % (num_runs, len(dags), total_tasks, times[0]))
+        msg += f"{times[0]:.4f}s"
+    print(msg)
```
Comment (uranusjr) [inline dev/perf/scheduler_dag_execution_timing.py]: Might be easier to do

```python
print(f"Time for ...", end="")
if len(times) > 1:
    print(f"...")
else:
    print(f"...")
```

File: dev/perf/sql_queries.py [hunk 6]
```diff
@@ -145,7 +145,7 @@ def make_report() -> list[Query]:
     for query in raw_queries:
         time, info, stack, sql = query.replace("@SQLALCHEMY ", "").split("|$")
         func, file, loc = info.split(":")
-        file_name = file.rpartition("/")[-1] if "/" in file else file
+        file_name = file.rpartition("/")[-1]
```
Comment (uranusjr) [inline dev/perf/sql_queries.py:148]:
```suggestion
        file_name = file.rsplit("/", 1)[-1]
```

A bit easier to understand? Probably not much difference so mostly just personal preference.

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Pull request merged**